### PR TITLE
Sample Document Processing Application

### PR DESCRIPTION
**Pull request by cbossie**
*Issue #, if available:*

*Description of changes:*
This adds a .NET 8.0 Serverless document processing and analysis application to the 2023 Sample Applications folder. The README has very explicit details as to what is included, but overall it include the following:
- .NET 8 Custom Runtime Lambda Functions
- Lambda Annotations (incl. main method creation)
- Lambda Powertools
- C# CDK application infrastructure as code.
- Textract (For query and Expense analysis)
- Step Functions (with Wait for Task Token pattern)
- Some unit tests for Textract model analysis.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by dhelper — CHANGES_REQUESTED** [review]
Need to update README.MD in 2023 root folder

File: SampleApplications/2023/ServerlessDocumentAnalysis/functions/src/ProcessTextractExpenseResults/Function.cs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        // Get the step functions Result
+        var textractExpenseModel = await _textractService.GetExpenses(processData.OutputBucket, processData.ExpenseOutputKey).ConfigureAwait(false);
+
+        // Get Expense Data and save to the DB
+        foreach (var id in textractExpenseModel.GetExpenseReportIndexes())
+        {
+            var report = new DocumentExpenseReport();
+
+            //Add Scalar Values            
+            foreach (var scalarSummaryField in textractExpenseModel.GetScalarSummaryFields(id))
```
Comment (dhelper) [inline SampleApplications/2023/ServerlessDocumentAnalysis/functions/src/ProcessTextractExpenseResults/Function.cs:44]: Consider extracting to method instead/in-addition to the "header" comments

File: SampleApplications/2023/ServerlessDocumentAnalysis/functions/src/ProcessingFunctions/Function.cs [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+    [LambdaFunction]
+    [Tracing]
+    [Metrics]
+    [Logging]
+    public async Task<IdMessage> InitializeHandler(S3StepFunctionCompositeEvent input, ILambdaContext _context)
+    {
+
+        //Initialize the Payload
+        ProcessData pl = new(_dataSvc.GenerateId())
```
Comment (dhelper) [inline SampleApplications/2023/ServerlessDocumentAnalysis/functions/src/ProcessingFunctions/Function.cs:82]: Consider extracting to methodcode block under comment to method

File: SampleApplications/2023/ServerlessDocumentAnalysis/infrastructure/.gitignore [hunk 3]
```diff
@@ -0,0 +1,350 @@
+# CDK asset staging directory
```
Comment (dhelper) [inline SampleApplications/2023/ServerlessDocumentAnalysis/infrastructure/.gitignore:1]: I think you can move the .gitignore file to the project root folder

File: SampleApplications/2023/ServerlessDocumentAnalysis/infrastructure/src/ServerlessDocAnalysis/StepFunctions/StepFunctionConstruct.cs [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+
+        // Log group for the step function
+        LogGroup stepFunctionLogGroup = new(this, "stepFunctionLogGroup", new Amazon.CDK.AWS.Logs.LogGroupProps
+        {
+            LogGroupName = GetLogGroupName("docProcessingWorkflow"),
+            RemovalPolicy = RemovalPolicy.DESTROY
+        });
+
+        // Update the step function body and use the passed in props
+        StateMachine = new StateMachine(this, $"{id}StateMachine", new StateMachineProps
```
Comment (dhelper) [inline SampleApplications/2023/ServerlessDocumentAnalysis/infrastructure/src/ServerlessDocAnalysis/StepFunctions/StepFunctionConstruct.cs:161]: Can extract different parts of build to methods and only leave on level of abstraction - https://www.code4it.dev/cleancodetips/coherent-levels-of-abstraction/

File: SampleApplications/2023/ServerlessDocumentAnalysis/functions/test/DocProcessingTest/TestTextractQuery.cs [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+public class TestTextractQuery
+{
+    private TextractDataModel? TextractData { get; set; }
+    private TextractAnalysisResult? TextractResult { get; set; }
+
+    [TestInitialize()]
+    public void Setup()
+    {
+        using FileStream jsonStream = File.OpenRead(@"TextractResults.json");
+        if (jsonStream is null)
```
Comment (dhelper) [inline SampleApplications/2023/ServerlessDocumentAnalysis/functions/test/DocProcessingTest/TestTextractQuery.cs:18]: Should neve have condition in tests, preferably use Asserts

File: SampleApplications/2023/ServerlessDocumentAnalysis/functions/test/DocProcessingTest/TestTextractQuery.cs [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    [TestMethod("Get Query Result")]
+    public void TestQueryResults()
+    {
+        var queryResultPatientName = TextractData?.GetQueryResults("patientname");
+
+        Assert.IsTrue(queryResultPatientName?.Count() == 2);
+
+        Assert.AreEqual(queryResultPatientName.Where(a => a.Text == "Edward Sang").Count(), 1);
```
Comment (dhelper) [inline SampleApplications/2023/ServerlessDocumentAnalysis/functions/test/DocProcessingTest/TestTextractQuery.cs:45]: Consider using multiple asserts where applicable (i.e. when testing the same result, different aspects): https://www.code4it.dev/cleancodetips/coherent-levels-of-abstraction/

**Review by dhelper — COMMENTED** [review]
Probably missing a few tests, at least unit tests for each lambda functions using mocks,

File: SampleApplications/2023/ServerlessDocumentAnalysis/infrastructure/.gitignore (hunk 3 above)
Comment (cbossie) [reply SampleApplications/2023/ServerlessDocumentAnalysis/infrastructure/.gitignore:1]: Good idea! Will make it happen.

File: SampleApplications/2023/ServerlessDocumentAnalysis/functions/test/DocProcessingTest/TestTextractQuery.cs (hunk 5 above)
Comment (cbossie) [reply SampleApplications/2023/ServerlessDocumentAnalysis/functions/test/DocProcessingTest/TestTextractQuery.cs:18]: Good call. Updating.

File: SampleApplications/2023/ServerlessDocumentAnalysis/functions/test/DocProcessingTest/TestTextractQuery.cs (hunk 6 above)
Comment (cbossie) [reply SampleApplications/2023/ServerlessDocumentAnalysis/functions/test/DocProcessingTest/TestTextractQuery.cs:45]: Understood. I simplified it and made what I was testing clearer. I didn't want to overcomplicate this one, for sure.

File: SampleApplications/2023/ServerlessDocumentAnalysis/infrastructure/src/ServerlessDocAnalysis/StepFunctions/StepFunctionConstruct.cs (hunk 4 above)
Comment (cbossie) [reply SampleApplications/2023/ServerlessDocumentAnalysis/infrastructure/src/ServerlessDocAnalysis/StepFunctions/StepFunctionConstruct.cs:161]: Abstracted the "build workflow chain structure" to another method. Left definition of different step / Log / EB resources in the constructer (to resemble typical CDK structure). Should make it look quite a bit simpler and easier to follow.

File: SampleApplications/2023/ServerlessDocumentAnalysis/functions/src/ProcessTextractExpenseResults/Function.cs (hunk 1 above)
Comment (cbossie) [reply SampleApplications/2023/ServerlessDocumentAnalysis/functions/src/ProcessTextractExpenseResults/Function.cs:44]: Refactored this significantly.

File: SampleApplications/2023/ServerlessDocumentAnalysis/functions/src/ProcessingFunctions/Function.cs (hunk 2 above)
Comment (cbossie) [reply SampleApplications/2023/ServerlessDocumentAnalysis/functions/src/ProcessingFunctions/Function.cs:82]: Refactored to simplify method and move business logic out of Lamdba function body.

**Comment by cbossie** [comment]
@dhelper I have updated based on your suggestions. Please take a look and let me know.

**Review by dhelper — APPROVED** [review]

**Comment by cbossie** [comment]
I added two small changes:
1) Update the test assets to have the right directory delimiter
2) Change the name of the "_context" parameter

**Pull request merged**